server/models/Collection.js

```diff
@@ -261,98 +241,56 @@ Collection.prototype.updateDocument = async function(updatedDocument) {
 Collection.prototype.moveDocument = async function(document, index) {
   if (!this.documentStructure) return;
 
-  const documentJson = await this.removeDocument(document, {
-    deleteDocument: false,
-  });
+  const documentJson = await this.removeDocumentInStructure(document);
   await this.addDocumentToStructure(document, index, { documentJson });
-
-  return this;
 };
 
-type DeleteDocumentOptions = {
-  deleteDocument: boolean,
+Collection.prototype.deleteDocument = async function(document) {
+  await this.removeDocumentInStructure(document, { save: true });
+  await document.deleteWithChildren();
 };
 
-/**
- * removeDocument is used for both deleting documents (deleteDocument: true)
- * and removing them temporarily from the structure while they are being moved
- * (deleteDocument: false).
- */
-Collection.prototype.removeDocument = async function(
+Collection.prototype.removeDocumentInStructure = async function(
   document,
-  options: DeleteDocumentOptions = { deleteDocument: true }
+  options?: { save?: boolean }
 ) {
   if (!this.documentStructure) return;
-
   let returnValue;
+  let unlock;
 
-  // documentStructure can only be updated by one request at the time
-  const unlock = await asyncLock('testLock');
-
-  const existingData = {
-    old: this.documentStructure,
-    documentId: document,
-    parentDocumentId: document.parentDocumentId,
-    options,
-  };
-
-  // Helper to destroy all child documents for a document
-  const deleteChildren = async documentId => {
-    const childDocuments = await Document.findAll({
-      where: { parentDocumentId: documentId },
-    });
-    childDocuments.forEach(async child => {
-      await deleteChildren(child.id);
-      await child.destroy();
-    });
-  };
+  if (options && options.save) {
+    // documentStructure can only be updated by one request at the time
+    unlock = await asyncLock(`collection-${this.id}`);
+  }
 
-  // Prune, and destroy if needed, from the document structure
-  const deleteFromChildren = async (children, id) => {
+  const removeFromChildren = async (children, id) => {
     children = await Promise.all(
       children.map(async childDocument => {
         return {
           ...childDocument,
-          children: await deleteFromChildren(childDocument.children, id),
+          children: await removeFromChildren(childDocument.children, id),
         };
       })
     );
 
-    const match = _.find(children, { id });
+    const match = find(children, { id });
     if (match) {
-      if (!options.deleteDocument && !returnValue) returnValue = match;
```

Comment:
Separated the code for managing the `documentStructure` and deleting / archiving individual docs